kernel/src/process/signal/mod.rs

```diff
@@ -223,30 +225,33 @@ pub fn handle_user_signal(
         .sig_context()
         .set(Some(ucontext_addr as Vaddr));
 
-    // 3. Set the address of the trampoline code.
+    // 3. Write the address of the restorer code.
     if flags.contains(SigActionFlags::SA_RESTORER) {
-        // If contains SA_RESTORER flag, trampoline code is provided by libc in restorer_addr.
-        // We just store restorer_addr on user stack to allow user code just to trampoline code.
+        // If the SA_RESTORER flag is present, the restorer code address is provided by the user.
         stack_pointer = write_u64_to_user_stack(stack_pointer, restorer_addr as u64)?;
-        trace!("After set restorer addr: user_rsp = 0x{:x}", stack_pointer);
+        trace!(
+            "After writing restorer addr: user_rsp = 0x{:x}",
+            stack_pointer
+        );
     } else {
-        // Otherwise we create a trampoline.
-        // FIXME: This may cause problems if we read old_context from rsp.
-        const TRAMPOLINE: &[u8] = &[
-            0xb8, 0x0f, 0x00, 0x00, 0x00, // mov eax, 15(syscall number of rt_sigreturn)
-            0x0f, 0x05, // syscall (call rt_sigreturn)
-            0x90, // nop (for alignment)
-        ];
-        stack_pointer -= TRAMPOLINE.len() as u64;
-        let trampoline_rip = stack_pointer;
-        user_space.write_bytes(stack_pointer as Vaddr, &mut VmReader::from(TRAMPOLINE))?;
-        stack_pointer = write_u64_to_user_stack(stack_pointer, trampoline_rip)?;
+        cfg_if::cfg_if! {
+            if #[cfg(target_arch = "x86_64")] {
+                // On x86-64, `SA_RESTORER` is mandatory and cannot be omitted.
+                // Ref: <https://elixir.bootlin.com/linux/v6.13/source/arch/x86/kernel/signal_64.c#L172>
+                return_errno_with_message!(Errno::EFAULT, "x86-64 should always use SA_RESTORER");
```

Comment:
For x86-64, I think it is too late to detect if `SA_RESTORER` is missing at this point. Since `SA_RESTORER` is mandatory, we should detect if it is missing on the `sigaction` or `signal` system calls. In this function, we should simply write the `restorer_addr` value to the bottom of the user stack. If it is null or invalid, then the user program will fault and then be killed by the kernel.

In short, we should harden the sanity check logic in the  `sigaction` or `signal` system calls.

Reply:
I have relocated the check to the rt_sigaction syscall. However, this implementation differs from Linux, so I have included a note to highlight this behavior.

```rust
// Note: Linux checks the SA_RESTORER flag 
// when initializing the signal stack,
// whereas we have moved this check forward 
// to prevent this action from being set.
// This may result in some differences from the behavior of Linux.
```

Additionally, I have enabled the `sigaction` test from gVisor to ensure that no tests are broken due to this change, and I have also fixed some existing issues.